### ISSUE TITLE
DrmSession$DrmSessionException with response code 400 after ad

### ISSUE THREAD

**Issue by Gobi1989** [google/ExoPlayer#3463]
Hi All,

**Issue description**
I'm using Exoplayer (r2.5.4) for playing DRM content along with Ads. As the player starts well with Ads and resume back to content video. We are getting an error while resume the DRM content video after 2nd/3rd AD.

**Reproduction steps**
Start the Drm content video along with Ads (pre Ads, 3 mid Ads and post Ads). After pre AD the content video starts to play as expected, now wait for second AD. The second Ad also will play as expected. The issue is after the second Ad (at some cases at after third AD) while resume back the content video the the Exoplayer throws an Exception as attached. 

**Link to test content**
Sent to dev.exoplayer@gmail.com

**Version of ExoPlayer being used**
com.google.android.exoplayer:exoplayer:r2.5.4
com.google.android.exoplayer:exoplayer-core:r2.5.4
com.google.android.exoplayer:exoplayer-dash:r2.5.4
com.google.android.exoplayer:exoplayer-ui:r2.5.4

**Device(s) and version(s) of Android being used**
Samsung S8, Android 7.0
Samsung GALAXY Note4, Android 6.0.1

**Note:-** The same value we are using in iOS Fair play and its working there. I had suppressed the AD and played the DRM content video its working fine at first time, i can able to play the whole video. While playing the same DRM content video again by clicking the play button at 2nd or 3rd try it fails with same exception. 

**Implmentation**
PlayerActivity was implemented as per link below
https://github.com/google/ExoPlayer/blob/release-v2/demo/src/main/java/com/google/android/exoplayer2/demo/PlayerActivity.java

**A full bug report captured from the device**
[ads_error_log.txt](https://github.com/google/ExoPlayer/files/1474146/ads_error_log.txt)

**Comment by Gobi1989** [comment]
Hi Team,
Please help in resolving this issue as it is a critical blocker issue for us.

Thanks in advance!

**Comment by Gobi1989** [comment]
Hi @edwinwong  

Please suggest me where we are wrong in these things. Kindly do the needful as soon as possible.

**Comment by andrewlewis** [comment]
It looks like a key request returned HTTP status code 400 (bad request). We will probably need a full bug report (not just a snippet of logcat) to help further.

I'm not aware of any issues doing ad insertion with DRM content. I tried playing one of the Widevine test streams from the demo app with ads inserted via the IMA extension, and it worked fine. Returning to content after playing an ad should work like loading content with an immediate seek. @ojw28 can you think of any potential issues specific to DRM playback with ads inserted that might explain this error?

**Comment by Gobi1989** [comment]
Hi @andrewlewis,

Even for me the content are playing well, only some particular content having the issue. 

Full bug report in the sense? (you meant full log cat?)

Thanks & Regards,
Gobi

**Comment by andrewlewis** [comment]
From the new issue template:

### A full bug report captured from the device
Capture a full bug report using "adb bugreport". Output from "adb logcat" or a
log snippet is NOT sufficient. Please attach the captured bug report as a file.
If you don't wish to post it publicly, please submit the issue, then email the
bug report to dev.exoplayer@gmail.com including the issue number in the subject
line.

**Comment by Gobi1989** [comment]
Hi @andrewlewis,

I have attached the bug report.
**Please let me know if this is not what you have expected**

[bugreport.zip](https://github.com/google/ExoPlayer/files/1487303/bugreport.zip)

**Comment by andrewlewis** [comment]
Please attach a bug report taken just after reproducing the error.

**Comment by Gobi1989** [comment]
Hi @andrewlewis,

I can't able to access the content now. Will share it shortly.

Regards,
Gobi

**Comment by Gobi1989** [comment]
Hi @andrewlewis,

Thanks for your patience. 

Herewith i have attached the bug report generated from command prompt and also device as well (Samsung GALAXY Note 4, Android 6.0.1).

It was taken just after i had the 400 exception.

[bugreport_from_cmd.txt](https://github.com/google/ExoPlayer/files/1494804/bugreport_from_cmd.txt)

[bugreport_from_mobile.txt](https://github.com/google/ExoPlayer/files/1494890/bugreport_from_mobile.txt)

Regards,
Gobi

**Comment by andrewlewis** [comment]
Neither of those files contains the `ExoPlaybackException` in the logcat attached to the original issue report?

**Comment by Gobi1989** [comment]
Hi @andrewlewis ,

I have taken the report as per this [link.](https://developer.android.com/studio/debug/bug-report.html)

Thanks & Regards,
Gobi

**Comment by andrewlewis** [comment]
I suggest plugging the device in, running `adb logcat`, reproducing the error, verifying that you can see the `ExoPlaybackException` in the output of `adb logcat`, then immediately running `adb bugreport`. That should ensure the bug report contains the error.

**Comment by Gobi1989** [comment]
Hi @andrewlewis ,

I have attached the **adb bugreport** and i had ensure that it has **ExoPlaybackException.**
Please let me know if you need any more inputs.

[bugreport.txt](https://github.com/google/ExoPlayer/files/1495598/bugreport.txt)

Thanks & Regards,
Gobi

**Comment by ojw28** [comment]
If the license server is responding with a 400 then it thinks something is wrong with the request. I think we'd really need to know what specifically it thinks is incorrect to investigate further.

Do you control the license server? If so, it should be fairly easy for you to add server-side debugging to work out exactly why 400s are being returned (e.g. because some specific field that the server expected in the request is missing, or similar). If not, you may have to ask whoever does provide the license server for help. Alternatively, if you could capture the exact content of one of the problematic requests and post it here, it may be possible for someone from Widevine to take a look for any obvious issues.

**Comment by BucherTomas** [comment]
Can't say for sure, but it seems that the client is calling the license server via GET method instead of POST method which Widevine proxies commonly require.

See in the log:

11-22 16:09:11.222  3139 15219 I Log     : URL = https://sdp-fng-webservices.quickplay.com/wvp/getlicense?[truncated_for_clarity]
11-22 16:09:11.232  3139 15219 I Log     : RequestMethod = GET

**Comment by Gobi1989** [comment]
Hi @ojw28 ,

We (myself) don't have any control over license server. I have captured the the **URL, Connection Request Properties, PostBody and some** over Log. I have shared the Log file over the mail with the
issue id as subject. From that log file you can see the problematic request that has Rersponse Code 400 at the bottom of the logs.

Regards,
Gobi

**Comment by Gobi1989** [comment]
Hi @BucherTomas ,

I have shared the Full Log file via mail, where you can see that the request fails after some iteration. In correct words the player had Ads integration and the player throws exception while resuming the content video after 2nd or 3rd Ads. 

Regards,
Gobi

**Comment by ojw28** [comment]
The server log is really confusing. It seems to mix logging for arbitrary many requests together, with no way to separate them back out so you can see what log lines correspond to which request. For example this snippet is (presumably) for two requests that are occurring in parallel:

```
URL = <redacted>
URL = <redacted>
ResponseHeaders = null
ResponseHeaders = null
Connection Request Properties = <redacted>
Connection Request Properties =<redacted>
UseCaches = true
UseCaches = true
RequestMethod = GET
RequestMethod = GET
...
```

* Can you find a way to provide the logging for requests coming from just the client that hit the error, split out by request?
* I'm not convinced the logged RequestMethod in the server logs is correct. It appears to always be set to GET, but the logs also show non-empty PostBody in some cases.

**Comment by Gobi1989** [comment]
Hi @ojw28 ,

Just shared the cleared log file with synchronous **request and response code** with subject REG| ISSUE 3466.

Our technical guy sentenced that **  **license URLs expire after a minute or so. I can't remember exactly how long the URL lives, but it's a relatively short time.  The player is supposed to retrieve the licence at the beginning and store it locally.**

Does it cause for 400 Exception?

If so how to restrict repeated call and maintain the license offline and use it in Exoplayer.

**Comment by ojw28** [comment]
It's the license server that's reporting the 400, so that's a question for whoever controls your license server (or someone who knows enough about it).

That said, it sounds likely. In general I'm not sure license servers should behave in this way. You probably don't want genuine offline licenses unless you're actually downloading the content. And if you're using streaming licenses, the client generally needs to be able to re-request the streaming license for as long as the user is allowed to watch the content, to handle cases such as the player needing to be restarted. How would such cases be handled otherwise? Also, what is the reason for making the URL expire so quickly? Something doesn't sound right about the way DRM is being implemented here...

**Comment by Gobi1989** [comment]
Hi @ojw28,

Finally they requested to call the license URL only once and store the license in offline and it will be used for playing content video.  

**How can i achieve these by setting mode as  DefaultDrmSessionManager.MODE_DOWNLOAD in DefaultDrmSessionManager ?**

**Comment by Gobi1989** [comment]
Hi @ojw28,

In our exoplayer we have included ima extension for Ads integration. The Ads were playing well but frequently the Exoplayer skips the Ads by **Unexpected CONTENT_RESUME_REQUESTED without stopAd**. We tried to figure out the steps to reproduce but we couldn't.

One more observation is if we forwarded the content before few seconds of Ads this issue happens but not for all the time.

Am not sure whether it is the right place to have discussion about Ads, if wrong please apologies.
Kindly provide the solution for this issue also.

**Comment by andrewlewis** [comment]
@Gobi1989 Please file a new issue for that and include steps to reproduce the issue. Then we can keep this issue focused on the DRM problem.

**Comment by Gobi1989** [comment]
Sure @andrewlewis

Any update about offline license management??

**Comment by Gobi1989** [comment]
Hi @andrewlewis @ojw28,

Can you please update the offline license management. As i'm not aware about it. Please suggest me with some snippet of code.

**Comment by ojw28** [comment]
Documentation for offline license management is tracked by https://github.com/google/ExoPlayer/issues/4100.
